### PR TITLE
First version

### PR DESCRIPTION
**Pull request by ceca2017**
First version for the library challenge

### CODE REVIEW

File: spec/library_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    allow(patron).to receive(:borrowed_books).and_return([])
+    allow(patron).to receive(:borrowed_books=)
+  end
+
+  it "has a collection of all books" do
+    expect(subject.collection).to be_instance_of(Array)
+  end
+
+  it "can show a list of books available" do
+    expect(subject.available_books[0][:available]).to eq true
```
Comment (diraulo) [inline spec/library_spec.rb:17]: This isn't really testing what the title says. Looks to me like you're just checking that the first book in your library is available. This test will fail as soon as the data file containing the list of books changes.
Give it some more thoughts :)

File: spec/library_spec.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  describe 'book check-out' do
+    it "is able to check out a book to a patron" do
+      item = { title: "Alfons och soldatpappan", author: "Gunilla Bergström" }
+      expected_output = { item: { title: "Alfons och soldatpappan",
+        author: "Gunilla Bergström" }, available: false,
+        return_date: (Date.today+30).to_s }
```
Comment (diraulo) [inline spec/library_spec.rb:25]: Isn't there a method on the `Date` class that will return you the date one month from now? 🤔 
I'd probably use that if there's any.

File: spec/patron_spec.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+require './lib/patron.rb'
+
+describe Patron do
+
+  arr_books = [{ item: { title: "Pippi Långstrump", author: "Astrid Lindgren" },
```
Comment (diraulo) [inline spec/patron_spec.rb:5]: This variable is only used once, to set yet another variable. You might as well just set `expected_output` with this content, no need for an intermediate variable unless you need it somewhere else.

File: lib/patron.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+class Patron
+
+  attr_accessor :borrowed_books
+
+  def initialize()
+    @borrowed_books = []
+  end
+
+  def borrow_book(attrs = {})
```
Comment (diraulo) [inline lib/patron.rb:11]: This seems like an action that belongs to the library? 🤔 
if you think about it for a sec... A patron would go to the library and ask to `borrow a book` (from the library) and not ask himself (the patron) for a book

File: lib/library.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  def set_return_date
+    Date.today + 30
+  end
+
+  def set_book_as_borrowed(book_index)
+    @collection[book_index][:available] = false
+    @collection[book_index][:return_date] = set_return_date.to_s
+  end
+
+  def find_book(item)
```
Comment (diraulo) [inline lib/library.rb:64]: Ruby has a method that does what you're trying to achieve here - Take a look at Enumerables in the ruby documentation and see if you can figure out which one 😉

File: lib/library.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    i = book[:index]
+    set_book_as_borrowed(i)
+    patron.borrowed_books << @collection[i]
+    update_book_collection
+    @collection[i]  # maybe not so nice?
+        #Line added so method will return a value to check against test.
+  end
+
+  def set_return_date
+    Date.today + 30
```
Comment (diraulo) [inline lib/library.rb:56]: There's a method on the Date class that will give you the date a month from now 😉

File: lib/library.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+  def load_book_collection
+    YAML.load_file('./lib/data.yml')
+  end
+
+  def update_book_collection
+    File.open('./lib/data.yml', 'w') { |f| f.write collection.to_yaml }
+  end
+
+  def perform_transaction(book, patron)
```
Comment (diraulo) [inline lib/library.rb:44]: I'd think of a better name for this method.. `perform_transaction` reminds me of a bank transaction (re: atm)

File: lib/library.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    when book[:available] == false
+      'Book is not available. It will be returned ' +
+        (@collection[book[:index]][:return_date]).to_s
+    else
+      perform_transaction(book, patron)
+    end
+  end
+
+  def available_books
+    avail_books = []
```
Comment (diraulo) [inline lib/library.rb:26]: You'll also find a method on the `Enumerable` class that does just what you're trying to achieve here

**Review by diraulo — COMMENTED** [review]
Overall good attempt 🙂

I like that you thought of adding a `README.md` to describe how to use your program 👍 

Have a look at the comments I left at various places in your code and see if you can fix some of that 😉

**Comment by ceca2017** [comment]
Thanks Raoul for your review and comments! I will check on them in more details. Thanks!

**Pull request closed without merging**